### PR TITLE
2561 make config api consistent

### PR DESCRIPTION
**Pull request by iilyak**

### CODE REVIEW

**Comment by kxepal** [comment]
Awesome work, but I'm in doubt is that reasonable to disallow atoms as Default value for `config:get`. By looking on referenced PRs you'd change a lot of `config:get` routines, and...let's pick this one from couchdb-couch:
```diff
-   case config:get(<<"httpd">>, <<"config_whitelist">>, null) of
-   null ->
+   case config:get("httpd", "config_whitelist", "null") of
+   "null" ->
         % No whitelist; allow all changes.
         handle_approved_config_req(Req, Persist);
         WhitelistValue ->
```

So, for old behaviour having `[httpd] config_whitelist = null` was not allowed. Now it's ok. So actually these changes introduced yet another state of deleted options in config INI file - when it matches the default in code (most defaults are nil and null). Atoms are good sentinel values for "no value" case and we should use them, not abuse.

File: test/config_tests.erl [hunk 1]
```diff
@@ -465,3 +358,92 @@ should_create_persistent_option() ->
...
+        ?assertMatch({{"httpd", "port", "80", false}, _}, wait_reply(Pid))
+    end).
+should_pass_correct_state_to_handle_config_change({Pid, _}) ->
+    ?_test(begin
+        ok = config:set("httpd", "port", "80", false),
+        ?assertMatch({_, {Pid, _, []}}, wait_reply(Pid)),
+        ok = config:set("update_state", "foo", "any", false),
+        ?assertMatch({Pid, _, ["foo"]}, wait_reply(Pid))
+    end).
+should_pass_correct_state_to_handle_config_stop({Pid, _}) ->
```
Comment (kxepal) [inline test/config_tests.erl]: These function needs a space between since they don't share the name and arity.

**Comment by iilyak** [comment]
The problem with `[httpd] config_whitelist = null` is very easy to solve using either.
```
case config:get(<<"httpd">>, <<"config_whitelist">>, undefined) of
    undefined ->
```
OR
```
case config:get(<<"httpd">>, <<"config_whitelist">>) of
    undefined ->
```

**Comment by kxepal** [comment]
@iilyak hm...true. Need to update those code after your PRs.

**Comment by iilyak** [comment]
I'll push new commit using `undefined`.

**Comment by kxepal** [comment]
@iilyak btw, it seems you'd missed ioq project. I made a change in my branch. Could you take a look on it? https://github.com/kxepal/couchdb-ioq/commit/b7b841294f7a37d9e9b59d2833e781e36bcad236

**Comment by kxepal** [comment]
@iilyak great! thank you!

**Comment by iilyak** [comment]
All PRs are updated. Summary of updates:
1 rename `handle_config_stop` into `handle_config_terminate`
2. make use of `undefined` default value when calling `config:get/3`
3. Add `handle_config_terminate(_, stop, _) -> ok;` clause to prevent event handler restart on termination

**Comment by kxepal** [comment]
All great!
@iilyak how do you think: if there need to use `config:get/3` with explicit default `undefined` value when the same value is used implicitly when you call `config:get/2`?

**Comment by iilyak** [comment]
I hope some day we can deprecate `config:get/2` so every caller would have to provide sensible default.

**Comment by kxepal** [comment]
Agreed.

**Comment by kxepal** [comment]
@iilyak all cool. one more thing before merge: could you please squash renaming and config default changes and misc fixing commits (indent in couchdb-couch)? We don't need to reflect changes in history that were instantly changed in the following commits (e.g. `nil -> "nil" -> undefined` when the only change that matters is `nil -> undefined` etc.).

**Comment by iilyak** [comment]
All dependent PRs are cleaned up and ready for merge

**Comment by kxepal** [comment]
All merged! Thank you! \o/

**Pull request merged**